sns/parity_math.c

```diff
@@ -796,6 +837,286 @@ static void xor_recover(struct m0_parity_math *math,
         }
 }
 
+#ifndef __KERNEL__
+static int isal_gen_recov_coeff_tbl(uint32_t data_count, uint32_t parity_count,
+				    struct m0_buf *failed_idx_buf,
+				    struct m0_buf *alive_idx_buf,
+				    uint8_t *encode_mat, uint8_t *g_tbls)
+{
+	uint8_t  *decode_mat = NULL;
+	uint8_t  *temp_mat = NULL;
+	uint8_t  *invert_mat = NULL;
+	uint32_t  unit_count;
+	uint32_t  i;
+	uint32_t  j;
+	uint32_t  r;
+	uint8_t	  s;
+	uint8_t	  idx;
+	int	  ret = 0;
+
+	M0_ENTRY();
+
+	unit_count = data_count + parity_count;
+
+	ALLOC_ARR_INFO(decode_mat, (unit_count * data_count), "decode matrix", ret);
+	if (ret != 0)
+		goto fini;
+
+	ALLOC_ARR_INFO(temp_mat, (unit_count * data_count), "temp matrix", ret);
+	if (ret != 0)
+		goto fini;
+
+	ALLOC_ARR_INFO(invert_mat, (unit_count * data_count), "invert matrix", ret);
+	if (ret != 0)
+		goto fini;
+
+	/* Construct temp_mat (matrix that encoded remaining frags)
+	 * by removing erased rows
+	 */
+	for (i = 0; i < data_count; i++) {
+		idx = ((uint8_t *)alive_idx_buf->b_addr)[i];
+		for (j = 0; j < data_count; j++)
+			temp_mat[data_count * i + j] =
+				encode_mat[data_count * idx + j];
+	}
+
+	/* Invert matrix to get recovery matrix */
+	ret = gf_invert_matrix(temp_mat, invert_mat, data_count);
+	if (ret != 0) {
+		ret = M0_ERR_INFO(ret, "failed to construct an %u x %u inverse "
+				  "of the input matrix", data_count, data_count);
+		goto fini;
+	}
+
+	/* Create decode matrix */
+	for (r = 0; r < failed_idx_buf->b_nob; r++) {
+		idx = ((uint8_t *)failed_idx_buf->b_addr)[r];
+		/* Get decode matrix with only wanted recovery rows */
+		if (idx < data_count) {    /* A src err */
+			for (i = 0; i < data_count; i++)
+				decode_mat[data_count * r + i] =
+					invert_mat[data_count * idx + i];
+		}
+		/* For non-src (parity) erasures need to multiply
+		 * encode matrix * invert
+		 */
+		else { /* A parity err */
+			for (i = 0; i < data_count; i++) {
+				s = 0;
+				for (j = 0; j < data_count; j++)
+					s ^= gf_mul(invert_mat[j * data_count + i],
+						    encode_mat[data_count * idx + j]);
+				decode_mat[data_count * r + i] = s;
+			}
+		}
+	}
+
+	ec_init_tables(data_count, (uint32_t)failed_idx_buf->b_nob, decode_mat, g_tbls);
+
+fini:
+	m0_free(temp_mat);
+	m0_free(invert_mat);
+	m0_free(decode_mat);
+
+	M0_LEAVE();
+
+	return ret;
+}
+
+static int sort_buffers(uint8_t **frags_in, uint8_t **frags_out,
```

Comment:
Two comments:
* Our function name is preferred to be: noun_verb(), like fails_sort().
* Please add Doxygen comments for this function: parameters, return values.

Reply:
Done.